### PR TITLE
Envsupport

### PR DESCRIPTION
**Pull request by callahad**
Addresses #32 

Still need to make it possible to supply the signing key via an env var, but I'd like to get this part reviewed and merged.

This is a depressingly small diff for the number of hours and false starts it took to get here. :/

### CODE REVIEW

File: config.toml.dist [hunk 1]
```diff
@@ -1,26 +1,53 @@
...
 listen_port = 3333
-base_url = "https://portier.example.com"
-token_validity = 600
 
-[[keys]]
-file = "private.pem"
+# Server's user-facing URL - Default: (none)
+public_url = 
+
+
```
Comment (stephank) [inline config.toml.dist]: Spacing is odd here. Think this'd look better if it didn't have newlines between individual fields, but just one between each section?

File: src/config.rs [hunk 2]
```diff
@@ -121,14 +93,77 @@ impl Default for Templates {
...
+                client_id: settings.client_id.clone(),
+                secret: settings.secret.clone(),
+                discovery_url: settings.discovery_url.clone(),
+                issuer_domain: settings.issuer_domain.clone(),
+            };
+
+            providers.insert(domain.to_string(), provider);
+        }
+
+
```
Comment (stephank) [inline src/config.rs]: Just one newline is fine here.

File: src/serde_types.in.rs [hunk 3]
```diff
@@ -3,53 +3,50 @@
...
-    pub discovery: String,
-    pub client_id: String,
-    pub secret: String,
-    pub issuer: String,
+#[derive(Clone,Debug,Deserialize)]
+struct TomlCryptoTable {
+    token_ttl: u16,
+    keyfiles: Vec<String>,
+}
+
```
Comment (stephank) [inline src/serde_types.in.rs]: Two newlines here

File: src/serde_types.in.rs [hunk 4]
```diff
@@ -3,53 +3,50 @@
...
+#[derive(Clone,Debug,Deserialize)]
+struct TomlRedisTable {
+    url: String,
+    session_ttl: u16,
+    cache_ttl: u16,
+    cache_max_doc_size: u16,
+}
+
```
Comment (stephank) [inline src/serde_types.in.rs]: Two newlines here

File: src/config.rs [hunk 5]
```diff
@@ -101,69 +101,118 @@ pub struct Provider {
...
-            providers.insert(domain.to_string(), provider);
-        }
+        self.redis_url = Some(toml_config.redis.url);
+
+        self.redis_session_ttl = toml_config.redis.session_ttl;
+
+        self.redis_cache_ttl = toml_config.redis.cache_ttl;
+
+        self.redis_cache_max_doc_size = toml_config.redis.cache_max_doc_size;
+
```
Comment (stephank) [inline src/config.rs]: I don't think all this spacing is really necessary here. I understand you want to keep it in struct order. Maybe group things by section then, with only a newline between each.

File: src/config.rs [hunk 6]
```diff
@@ -101,69 +101,118 @@ pub struct Provider {
...
+            smtp_server: self.smtp_server.unwrap(),
+            providers: self.providers,
             templates: Templates::default(),
-        })
+        }
     }
 }
+
+
+pub struct Config {
```
Comment (stephank) [inline src/config.rs]: This should be above Builder, so Config structs are grouped together. I think it'll also make more sense in the diff, then.

File: src/main.rs [hunk 7]
```diff
@@ -52,9 +52,11 @@ fn main() {
...
 
     // Read the configuration from the provided file.
-    let app = Arc::new(
-        broker::Config::from_toml_file(&args.arg_CONFIG).unwrap()
-    );
+    let mut builder = broker::config::Builder::new();
+    builder.update_from_file(&args.arg_CONFIG);
+    let config = builder.done();
+
+    let app = Arc::new(config);
```
Comment (stephank) [inline src/main.rs]: `let app = Arc::new(builder.done())`

File: src/config.rs [hunk 8]
```diff
@@ -145,31 +145,55 @@ impl Builder {
         let toml_config: TomlConfig = toml::decode_str(&file_contents)
             .expect("unable to parse config file");
 
-        self.listen_ip = toml_config.server.listen_ip;
+        if let Some(table) = toml_config.server {
+            if let Some(val) = table.listen_ip {
+                self.listen_ip = val;
+            }
```
Comment (stephank) [inline src/config.rs]: Here, again, less whitespace would probably increase readability. The braces already create plenty of air.

File: src/serde_types.in.rs [hunk 9]
```diff
@@ -3,6 +3,7 @@
 // See https://serde.rs/codegen-stable.html for more information.
 
 
+/// Intermediate structure for deserializing TOML files
```
Comment (stephank) [inline src/serde_types.in.rs]: Should be in the commit that introduced this struct.

File: src/config.rs [hunk 10]
```diff
@@ -195,11 +229,43 @@ impl Builder {
...
+
+                if values.issuer_domain.is_some() {
+                    let provider = self.providers.entry(domain.clone())
+                        .or_insert_with(|| ProviderBuilder::new());
+
+                    provider.issuer_domain = values.issuer_domain;
+                }
+            }
+        }
+
```
Comment (stephank) [inline src/config.rs]: Again, reduce whitespace.

File: src/serde_types.in.rs [hunk 11]
```diff
@@ -51,3 +51,69 @@ struct TomlProviderTable {
     discovery_url: Option<String>,
     issuer_domain: Option<String>,
 }
+
```
Comment (stephank) [inline src/serde_types.in.rs]: Two newlines.

File: src/serde_types.in.rs [hunk 12]
```diff
@@ -51,3 +51,69 @@ struct TomlProviderTable {
...
+    broker_from_name: Option<String>,
+    broker_from_address: Option<String>,
+    broker_smtp_server: Option<String>,
+    broker_gmail_client: Option<String>,
+    broker_gmail_secret: Option<String>,
+    broker_gmail_discovery: Option<String>,
+    broker_gmail_issuer: Option<String>,
+}
+
+impl EnvConfig {
```
Comment (stephank) [inline src/serde_types.in.rs]: Should be in `config.rs`, I believe?

File: src/serde_types.in.rs [hunk 13]
```diff
@@ -51,3 +51,69 @@ struct TomlProviderTable {
...
+            broker_from_address: env::var("BROKER_FROM_ADDRESS").ok(),
+
+            broker_smtp_server: env::var("BROKER_SMTP_SERVER").ok(),
+
+            broker_gmail_client: env::var("BROKER_GMAIL_CLIENT").ok(),
+
+            broker_gmail_secret: env::var("BROKER_GMAIL_SECRET").ok(),
+
+            broker_gmail_discovery: env::var("BROKER_GMAIL_DISCOVERY").ok(),
+
```
Comment (stephank) [inline src/serde_types.in.rs]: Less whitespace.

File: src/config.rs [hunk 14]
```diff
@@ -264,6 +265,92 @@ impl Builder {
...
+            provider.discovery_url = Some(val);
+        }
+
+        if let Some(val) = env_config.broker_gmail_issuer {
+            let provider = self.providers.entry("gmail.com".to_string())
+                .or_insert_with(|| ProviderBuilder::new());
+
+            provider.issuer_domain = Some(val);
+        }
+
```
Comment (stephank) [inline src/config.rs]: Less whitespace.

File: src/config.rs [hunk 15]
```diff
@@ -264,6 +265,92 @@ impl Builder {
...
+            self.public_url = Some(val);
+        }
+
+        if let Some(val) = env_config.broker_token_ttl {
+            self.token_ttl = val;
+        }
+
+        if let Some(val) = env_config.broker_keyfiles {
+            // Should this append instead of replacing?
+            // Append seems more convenient, but everything else replaces.
```
Comment (stephank) [inline src/config.rs]: I'd say replacing makes the most sense here. You can remove this comment, IMO.

File: src/config.rs [hunk 16]
```diff
@@ -137,6 +138,15 @@ impl ProviderBuilder {
         }
     }
 
+    pub fn new_gmail() -> ProviderBuilder {
```
Comment (stephank) [inline src/config.rs]: Just make all of these changes a separate commit, before the README update.

File: src/crypto.rs [hunk 17]
```diff
@@ -26,19 +26,26 @@ pub struct NamedKey {
...
-        if key_res.is_err() {
-            return Err("could not instantiate private key");
+        match PKey::private_key_from_pem(&mut BufReader::new(private_key_file)) {
+            Err(_) => Err("could not instantiate private key"),
+            Ok(pkey) => {
+                let mut hasher = hash::Hasher::new(hash::Type::SHA256);
+                hasher.write(pkey.get_rsa().e().unwrap().to_vec().as_slice()).unwrap();
+                hasher.write(b".").unwrap();
+                hasher.write(pkey.get_rsa().n().unwrap().to_vec().as_slice()).unwrap();
+
```
Comment (stephank) [inline src/crypto.rs]: This newline can go. :)

File: src/config.rs [hunk 18]
```diff
@@ -101,69 +101,118 @@ pub struct Provider {
 }
 
 
-pub struct Config {
+pub struct Builder {
```
Comment (stephank) [inline src/config.rs]: Maybe `ConfigBuilder`? Would mirror `ProviderBuilder` and friends down the road.

**Comment by callahad** [comment]
This is why we need something like rustfmt ;)

TODO to address review comments:

- [ ] Rename `Builder` to `ConfigBuilder`. Initially named as such because I though we could get away without needing a separate ProviderBuilder
- [ ] Move the `impl EnvConfg` out of `serde_types.in.rs` and into `config.rs`
- [ ] Clean up whitespace

Working on that now.

File: src/email.rs [hunk 19]
```diff
@@ -6,7 +6,7 @@ use super::error::{BrokerError, BrokerResult};
 use super::lettre::email::EmailBuilder;
 use super::lettre::transport::EmailTransport;
 use super::lettre::transport::smtp::SmtpTransportBuilder;
-use super::{AppConfig, create_jwt};
+use super::{Config,create_jwt};
```
Comment (djc) [inline src/email.rs]: Nit: space after comma, please.

File: README.rst [hunk 20]
```diff
@@ -27,84 +27,36 @@ This should fetch and install the Portier broker. The binary is installed into
...
+.. code-block:: shell
 
-**token_validity** is a value in seconds, that determines how long outgoing
-authentication tokens are allowed to live. Defaults to 600s, or 10 minutes.
+   $ portier-broker config.toml
 
-**keys** is a list of RSA private keys, with a ``file`` path for each key.
-Multiple keys can be used to implement key rotation. By default, the last key
-in the list will be used for signing the outgoing JWTs.
+You will also need a Redis server and an outgoing SMTP server.
```
Comment (djc) [inline README.rst]: I think there's too much going on in this commit, making it hard to carefully review. The move from config-in-README to docs-in-TOML should probably be one commit on its own.

File: src/config.rs [hunk 21]
```diff
@@ -121,14 +93,77 @@ impl Default for Templates {
...
 
+pub struct Provider {
+    pub client_id: String,
+    pub secret: String,
+    pub discovery_url: String,
+    pub issuer_domain: String,
+}
+
+
+pub struct Config {
```
Comment (djc) [inline src/config.rs]: Renaming `AppConfig` to `Config` should be a commit on its own, too, though I'm not sure what the goal of this is?

File: src/email.rs [hunk 22]
```diff
@@ -57,7 +57,7 @@ pub fn request(app: &AppConfig, email_addr: EmailAddress, client_id: &str, nonce
 
     // Generate the URL used to verify email address ownership.
     let href = format!("{}/confirm?session={}&code={}",
-                       app.base_url,
+                       app.public_url,
```
Comment (djc) [inline src/email.rs]: All these config field name changes can then also go in a separate commit, I think, but again, many of these feel somewhat gratuitous.

File: src/config.rs (hunk 5 above)
Comment (djc) [inline src/config.rs]: +1

File: src/lib.rs [hunk 23]
```diff
@@ -33,7 +33,7 @@ use urlencoded::{UrlEncodedBody, UrlEncodedQuery};
 
 pub mod error;
 pub mod config;
-pub use config::Config;
+pub use config::{Builder,Config};
```
Comment (djc) [inline src/lib.rs]: Nit: space after comma.

File: src/serde_types.in.rs (hunk 13 above)
Comment (djc) [inline src/serde_types.in.rs]: Does it really make sense to repeat the `broker_` in the property names here?

File: src/config.rs [hunk 24]
```diff
@@ -114,7 +114,41 @@ pub struct Builder {
     pub from_name: String,
     pub from_address: Option<String>,
     pub smtp_server: Option<String>,
-    pub providers: HashMap<String, Provider>,
+    pub providers: HashMap<String, ProviderBuilder>,
```
Comment (djc) [inline src/config.rs]: I don't quite get the goal of this stuff. Why do we need the incrementalism? Would make for a good addition to the commit message, too.

**Comment by callahad** [comment]
Closing in favor of #50
